cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -299,51 +301,95 @@ func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathM
+	delay := 30 * time.Second
+	retries := 10
+	retryInterval := 15 * time.Second
+	if _, err = newEtcdClient.WaitForStatus(delay, retries, retryInterval); err != nil {
+		// Despite the fact that upgradeComponent was successful, there is something wrong with etcd cluster
+		// First step is to restore back up of datastore
+		if err := rollbackEtcdData(cfg, fmt.Errorf("etcd cluster is not healthy after upgrade: %v rolling back", err), pathMgr); err != nil {
+			// Even copying back datastore failed, no options for recovery left, bailing out
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
+		}
+		// Old datastore has been copied, rolling back old manifests
+		if err := rollbackOldManifests(recoverManifests, err, pathMgr, true); err != nil {
+			// Rolling back to old manifests failed, no options for recovery left, bailing out
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
 		}
+		// Since rollback of the old etcd manifest was successful, checking again the status of etcd cluster
+		if _, err := oldEtcdClient.GetStatus(); err != nil {
+			// Nothing else left to try to recover etcd cluster
+			return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, the backup of etcd database is stored here:(%s)", err, backupEtcdDir)
+		}
+
+		return true, fmt.Errorf("fatal error upgrading local etcd cluster: %v, rolled the state back to pre-upgrade state", err)
 	}
 
 	return false, nil
 }
 
 // StaticPodControlPlane upgrades a static pod-hosted control plane
-func StaticPodControlPlane(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, etcdUpgrade bool, oldEtcdCluster, newEtcdCluster etcdutil.Cluster) error {
+func StaticPodControlPlane(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, etcdUpgrade bool, oldEtcdClient, newEtcdClient etcdutil.Client) error {
 	recoverManifests := map[string]string{}
 	var isTLSUpgrade bool
+	var isExternalEtcd bool
 
 	beforePodHashMap, err := waiter.WaitForStaticPodControlPlaneHashes(cfg.NodeName)
 	if err != nil {
 		return err
 	}
 
-	// etcd upgrade is done prior to other control plane components
-	if etcdUpgrade {
-		previousEtcdHasTLS, err := oldEtcdCluster.HasTLS()
-		if err != nil {
-			return fmt.Errorf("failed to determine if previous etcd was using TLS: %v", err)
+	if oldEtcdClient == nil {
+		if len(cfg.Etcd.Endpoints) > 0 {
+			// External etcd
+			isExternalEtcd = true
+			client, err := etcdutil.NewClient(
+				cfg.Etcd.Endpoints,
+				cfg.Etcd.CAFile,
+				cfg.Etcd.CertFile,
+				cfg.Etcd.KeyFile,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to create etcd client for external etcd: %v", err)
+			}
+			oldEtcdClient = client
+			// Since etcd is managed externally, the new etcd client will be the same as the old client
```

Comment:
I like that this structure is explicit because it makes interesting upgrade options possible for `externalEtcd -> staticPodEtcd` and vice versa.